qiskit/circuit/commutation_checker.py

```diff
@@ -93,84 +77,291 @@ def commute(
-            ):
-                return False
-
-        # The main code is adapted from commutative analysis.
-        # Assign indices to each of the qubits such that all `node1`'s qubits come first, followed by
-        # any _additional_ qubits `node2` addresses.  This helps later when we need to compose one
-        # operator with the other, since we can easily expand `node1` with a suitable identity.
-        qarg = {q: i for i, q in enumerate(qargs1)}
-        num_qubits = len(qarg)
-        for q in qargs2:
-            if q not in qarg:
-                qarg[q] = num_qubits
-                num_qubits += 1
-        qarg1 = tuple(qarg[q] for q in qargs1)
-        qarg2 = tuple(qarg[q] for q in qargs2)
-
-        node1_key = (op1.name, self._hashable_parameters(op1.params), qarg1)
-        node2_key = (op2.name, self._hashable_parameters(op2.params), qarg2)
-        try:
-            # We only need to try one orientation of the keys, since if we've seen the compound key
-            # before, we've set it in both orientations.
-            return self.cache[node1_key, node2_key]
-        except KeyError:
-            pass
-
-        operator_1 = Operator(op1, input_dims=(2,) * len(qarg1), output_dims=(2,) * len(qarg1))
-        operator_2 = Operator(op2, input_dims=(2,) * len(qarg2), output_dims=(2,) * len(qarg2))
-
-        if qarg1 == qarg2:
-            # Use full composition if possible to get the fastest matmul paths.
-            op12 = operator_1.compose(operator_2)
-            op21 = operator_2.compose(operator_1)
+        structural_commutation = _commutation_precheck(
+            op1, qargs1, cargs1, op2, qargs2, cargs2, max_num_qubits
+        )
+
+        if structural_commutation is not None:
+            return structural_commutation
+
+        commutation_lookup = self.check_commutation_entries(
+            op1, qargs1, cargs1, op2, qargs2, cargs2
+        )
+
+        if commutation_lookup is not None:
+            return commutation_lookup
+
+        # Compute commutation via matrix multiplication
+        is_commuting = _commute_matmul(op1, qargs1, op2, qargs2)
+
+        # Store result in this session's commutation_library
+        # TODO implement LRU cache or similar
+        # Rebuild cache if current cache exceeded max size
+        if self._current_cache_entries >= self._cache_max_entries:
+            self._cached_commutations = {}
+
+        first_op_tuple, second_op_tuple = _order_operations(
+            op1, qargs1, cargs1, op2, qargs2, cargs2
+        )
+        first_op, first_qargs, _ = first_op_tuple
+        second_op, second_qargs, _ = second_op_tuple
+        first_params = first_op.params
+        second_params = second_op.params
+
+        self._cached_commutations.setdefault((first_op.name, second_op.name), {}).setdefault(
+            _get_relative_placement(first_qargs, second_qargs), {}
+        )[(_hashable_parameters(first_params), _hashable_parameters(second_params))] = is_commuting
```

Comment:
In general, the changes in this file look good. Can you please add a comment and hopefully with an explicit example showing how gates with hashable parameters would be represented in the commutativity dict. For instance, let's create multiple `LinearFunction`s from different matrices, with some pairs commuting and some pairs not commuting, how would the commutativity dict entry look like?

Reply:
The entry would be 
> dict['linear_function', 'linear_function'][relative qubit placement of LinearFunction gates][bytes representing parameter matrix] := True/False